src/components/common/dialog.tsx

```diff
@@ -0,0 +1,910 @@
+        if (from_path in this.validation_state) {
+            console.log("RENAME", from_path, to_path);
+            this.validation_state[to_path] = this.validation_state[from_path];
+            delete this.validation_state[from_path];
+        }
+        for (const k in this.validation_state) {
+            if (k.indexOf(from_path + ".") == 0) {
+                const to = to_path + k.substring(from_path.length);
+                console.log("RENAME.", k, to);
+                this.validation_state[to] = this.validation_state[k];
+                delete this.validation_state[k];
+            }
+        }
+    }
+
+    // For buttons
+
+    async run_action(func: (vals: V) => Promise<void>) {
+        this.error = null;
+        this.busy = true;
+        this.update();
+        if (!await this.validate()) {
+            this.busy = false;
+            this.update();
+            return;
+        }
+
+        try {
+            cockpit.assert(this.values);
+            await func(this.values);
+        } catch (ex) {
+            console.error(String(ex));
+            this.error = ex;
+        }
+
+        this.busy = false;
+        this.update();
+    }
+
+    // PUBLIC
+
+    constructor(init: V, validate: (dlg: DialogState<V>) => void) {
+        super();
+        this.validate_callback = validate;
+        this.values = init;
+    }
+
+    value<K extends keyof V>(tag: K, update_func?: ((val: V[K]) => void) | undefined): DialogValue<V[K]> {
+        return new DialogValue<V[K]>(
+            this as DialogState<unknown>,
+            () => this.values![tag],
+            (val) => {
+                console.log("TOP SET", tag, JSON.stringify(val));
+                this.values = { ...this.values, [tag]: val };
+                this.update();
+                if (this.online_validation)
+                    this.trigger_validation();
+                if (update_func)
+                    update_func(val);
+            },
+            String(tag));
+    }
+
+    async validate(): Promise<boolean> {
+        this.trigger_validation();
+        // TODO - wait for the asyncs to be done
+        return Object.keys(this.validation).length == 0;
+    }
+}
+
+export function useDialogState<V extends object>(
+    init: V | (() => V),
+    validate: (dlg: DialogState<V>) => void
+) : DialogState<V> {
+    const dlg = useObject(() => {
+        if (typeof init == "function")
+            return new DialogState(init(), validate);
+        else
+            return new DialogState(init, validate);
+    }, null, []);
+    useOn(dlg, "changed");
+    return dlg;
+}
+
+// export function useDialogState_async<V>(...)
+
+// Common elements
+
+export class DialogError {
+    title: string;
+    details: React.ReactNode;
+
+    constructor(title: string, details: React.ReactNode) {
+        this.title = title;
+        this.details = details;
+    }
+
+    toString() {
+        return this.title + ": " + String(this.details);
```

Comment:
These 2 added lines are not executed by any test.  [Details](https://logs-cockpit.apps.ocp.cloud.ci.centos.org/pull-2435-e22b04b7-20251204-081542-fedora-43-devel/Coverage/lcov/github-pr.diff.gcov.html)